Code review request for Luca Crema and Riccardo Crociani

ricdezen opened the pull request:
I'm in need of a code review for the following 4 files:

- NetworkDictionary.java
- SmsHandler.java
- SmsReceiver.java
- SMSManager.java

Where referenced, Peer and Message refer to the Peer and Message interfaces defined by the whole class, Resource slightly differs.
Each class contains a comment regarding its context.
The interfaces Resource.java and Dictionary.java are included for a better view of the global context.
Signed-off-by: Riccardo De Zen <riccardo.dezen@studenti.unipd.it>

Commit:
Marked relevant files for code review.

Signed-off-by: Riccardo De Zen <riccardo.dezen@studenti.unipd.it>

CremaLuca commented on networkmanager/src/main/java/com/dezen/riccardo/networkmanager/NetworkDictionary.java:
```diff
@@ -9,7 +9,10 @@
+ * involving relationships between Peers and Resources.
+ * Class implementing Dictionary. Conceived as a double dictionary on SMSPeer and StringResource,
+ * also allowing Peer ownership of Resources.
  * Due to trouble with testing android class. Lists have been used.
```
Quali problemi?

ricdezen replied:
La classe ArraySet non può venire testata con gli UnitTest poichè facendo parte di un pacchetto della libreria Android ha bisogno dei mock test, tuttavia il problema non viene segnalato in compilazione se un suo esemplare è contenuto in un'altra classe, come Dictionary. Una volta identificato il problema l'implementazione era già proseguita in altro modo.

CremaLuca reviewed (changes requested):
**SMSManager**

Generale:
- Nome incostante con il resto delle classi (Sms/SMS)? Lo hai fatto per allinearti agli altri? Non preoccuparti anche se due lettere sono minuscole, lo scopo di avere un interfaccia uguale è quello di non dover cambiare parametri, ordine dei parametri e nome delle classi se si volesse cambiare libreria. Scegli uno dei due e rimani consistente con la tua scelta.

Variabili:
- r/s/d Listener, potevi sprecarti un pochino di più per dare il nome (receivedListener, sentListener, deliveredListener).
OPINIONE PERSONALE:
- Context "owner"? È solo un context, non dipende esclusivamente da quello, io cambierei nome in qualcosa tipo currentContext o ctx.

Metodi:
- getInstance
  - context viene assegnato solo se instance=null, e se cambiasse il context ad istanza ancora attiva?

- drop
  - drop cosa? SmsManager.drop() non riporta alla mente ciò che effettivamente fa, prova a rinominarlo in dropContext().
  - Perchè rimuove anche il listener? Non potrebbe essere una classe ancora valida anche se il context muore?

- costruttore
  - molto brutto averlo come terzo metodo, di solito si tende a lasciarlo come il primo.
  - buona l'eccezione, potevi specificarlo nelle precondizioni ma così sei in una botte di ferro.

- sendMessage
  - mancano le specifiche, suppongo le abbia ereditate da CommunicationHandler ma quello che fa il metodo è specifico per gli SMS quindi avrei speso due parole.

-addReceiveListener
  - sono abbastanza perplesso sul fatto che vadano scritte tre righe di codice per fare una sola cosa, credo che il problema sia nell'handler.


**SmsHandler**

Generale:
- Come ho detto prima, inconstistente con il nome rispetto al manager
- Come mai hai usato dei flag per indicare se chiamare onSend e onDelivery? Non puoi semplicemente avere sempre il broadcastReceiver che intercetta e chiama il listener solo se null?

Variabili:
- tutte le costanti cominciano con SMS_HANDLER? Perchè non togliere SMS_HANDLER e aggiungere più informazioni dopo, ad esempio SMS_HANDLER_RECEIVED_BROADCAST suppongo sia qualcosa di più tipo RECEIVED_BROADCAST_INTENT_NAME
- "OnSmsEventListener listener", nome più generico non potevi scegliere, specifica meglio nel nome che tipo di listener è.

Metodi:
- costruttore
  - tutto ok, listener/scAddress = null che non serve e altre assegnazioni che potevi fare nella dichiarazione delle variabili (es i boolean = false).

- sendSMS
  - sembra ok, anche se io avrei fatto che il listener per i messaggi inviati fosse possibile settarlo diverso per ogni messaggio però sono scelte (nella tua scelta se cambio listener mentre il messaggio si sta inviando viene chiamato il listener sbagliato).

- registerReceiver(Context)
  - TODO? Mi dai da revisionare classi incomplete?
  - questo registra solo il listener per messaggio ricevuto, non lo potevi rinominare in registerReceviedReceiver o qualcosa del genere?

- registerReceiver(Context, bool, bool, bool)
  - altro TODO nelle specifiche.
  - manca il @param received nelle specifiche.
  - usi lo stesso BroadcastReceiver per tutte e tre le azioni? Non sarebbe meglio avere un Receiver diverso almeno per ricezione e (invio & consegna)? Sono due operazioni completamente diverse, metterle insieme mi sembra che renda pesante il BroadcastReceiver (che al momento devo ancora leggere).

- setListener
  - consiglio per disambiguazione di chiamare il nuovo listener qualcosa come "newListener" altrimenti si confonde con la variabile globale, oppure rinomini quella globale come scritto sopra

- shouldHandleIncomingSms
  - è un metodo pubblico perchè va chiamato dalla applicazione?

- fetchUnreadMessages
  - altro TODO
  - come mai la classe che gestisce l'invio e la ricezione dei messaggi deve occuparsi anche del database? Delega questo tipo di cose ad un altra classe dedicata al database, non tenere tutto unito.
  - come mai è un metodo pubblico? Va chiamato dalla applicazione?
  - sei sicuro che una funzione del genere (database) vada implementata dentro a SmsHandler? Non potrebbe esserci una classe intermedia che viene chiamata alla ricezione dei messaggi e che legge e scrive nel database? È obbligatorio per tutte le applicazioni avere un database dei messaggi che arrivano? Anche perchè ad essere sincero non ho capito chi scriva nel DB e dove.


**OnSmsEventListener**

Generale:
- obblighi tutte le classi a gestire sia l'invio che la ricezione? Solitamente voglio avere una classe che implementa questa interfaccia che gestisca solo una delle due cose.
- resultCode come lo interpreto? Come faccio a sapere se il messaggio è stato inviato? Su quali valori varia? Potevi usare un Enum personalizzato al suo posto?


**SmsEventReceiver** 

Generale:
- ha un singolo metodo enorme, non potevi delegare la ricostruzione del messaggio inviato/consegnato ad un metodo "SMSMessage getMessage(String peer, String message)"?
- hardwiring degli intent.getExtras "address" e "message"
- in tutti e tre gli if controlli se listener != null, non puoi tirarlo fuori?
- consiglio: quando hai tanti controlli di validità, al posto di fare delle grandi parentesi che innestate una dentro all'altra fai qualcosa tipo: if(listener == null) return; così non devi spostarti a destra ad ogni controllo
- extra: il controllo se il messaggio appartiene alla applicazione lo fai con il contains(), e se un giorno volessi cambiare formato? Io estrarrei il controllo e lo metterei da qualche parte come metodo statico

**SmsReceiver**

Generale:
- altri TODO

Metodi:
- onReceive
  - bene che hai delegato il filter()
  - anche qui ci starebbe bene: if(messages.size() <= 0) return;
  - evita Log.d "Forwarding intent...", sono messaggi 'verbosi', quindi se vuoi tenerlo per debug inseriscilo in Log.v
  - anche qui gestisci dentro alla classe la creazione/recupero del database, delega funzioni extra il più possibile

- filter
  - bene private
  - consiglio: come prima, il controllo sulla appartenenza del messaggio alla applicazione potrebbe essere estrapolato e messo da un altra parte

- sendImplicitBroadcast
  - male public
  - stai facendo un broadcast interno alla libreria per dire che è arrivato un messaggio. Hai mai pensato di passare ad un listener? Oppure direttamente ad avere il broadcastReceiver che chiama il listener che la applicazione ha passato all'handler senza passare per l'handler?

**Dictionary**

Generale:

- un altro TODO
- non c'è il metodo getResource(ResourceKey)? Uno deve sempre prendere tutto l'array di risorse per sfogliarlo? Se la risposta è no potrebbe essere comodo passare ad un map<K,V> altrimenti può rimanere così

**NetworkDictionary**

Generale:
- trouble with testing? Quali problemi?
- mi piace che hai dato nomi corretti nei parametri come "newPeer" e "peerToRemove"
- Ho controllato tutti i metodi e mi sembrano ottimi, niente da segnalare

Test

- Non ci sono i test! Non mi sono stati dati da revisionare.
EDIT: I test ci sono e sono molti, solo che non sono stati presi da revisionare


**CheckList utilizzata:**
- Nomenclatura
  - Usa variabili con nomi coerenti?
  - I nomi dei metodi cominciano con il verbo?
  - Struttura del codice
  - È presente hardwiring?
  - Ci sono ancora dei TODO/FIXME?
  - Potrebbe utilizzare un design pattern noto?
  - Ci sono commenti superflui?
  - Le eccezioni sono gestite correttamente?
  - I metodi fanno tutti cose coerenti con la classe?
  - Si possono spezzare ulteriormente i metodi?
- Interfaccia esterna
  - È incapsulato correttamente?
  - È comodo per l’utilizzatore?
  - Richiede l’utilizzo di tanti altri package esterni?
  - È versatile/può essere usato in più situazioni simili?
- Specifiche
  - Ci sono tutte?
  - Sono troppo permissive/stringenti?
  - La classe ha descrizione e @author?
- Test
  - Ci sono?
  - I nomi dei metodi seguono una struttura coerente e sono autoesplicativi?
  - Si potrebbe utilizzare un metodo @Before?

RicDeZen commented:
Le uniche cose che non riconosco pienamente come colpe sono alcuni dei punti di domanda:
- SMSManager.drop() (Ora rinominato in onContextDestroyed() ) rimuove anche il listener poichè serve a indicare che il receiver interno ad SmsHandler non ha più nessun ascoltatore e quindi non ha più ragione di esistere appesantendo il sistema.
- SMSReceiver.sendImplicitBroadcast() (che in effetti va messo private) serve per triggerare un receiver che sia presente in un altro modulo e che non conosco. Il metodo è complesso ma è l'unica soluzione che ho trovato attualmente per Android 8.0 e oltre.

Per il resto mi scuso per il disastro che è il codice, del 90% di questi problemi ero a conoscenza, ma li ho sottovalutati, ho voluto fare troppe cose troppo in fretta e ora sto facendo molta fatica a sistemare.

RicDeZen commented:
Altri appunti riguardo le scelte di implementazione che ho trovato durante la correzione:
- SMSManager.getInstance() assegna il context solo quando l'istanza viene creata la prima volta ed è intenzionale, essendovi un receiver registrato, anche se interno, è buona norma secondo la documentazione Android che vada deregistrato nel momento in cui il context viene distrutto. Chiamare unregisterReceiver da un altro Context rispetto a quello di registrazione non rimuove il suddetto Receiver.
- Avere i tre eventi, anche se abbastanza distinti in un unico Receiver è intenzionale anch'esso, poichè la documentazione Android suggerisce di ridurre quanto più possibile il numero di Receiver, senza fare alcuna menzione riguardo il volume dei singoli, è anche vero che non è ottimale per via della natura dei PendingIntent, ma non abbiamo trovato una soluzione migliore ancora.
- Il resultCode per la conferma di invio e ricezione è quello di default della classe Activity, il PendingIntent genera un valore intero per tale result code, e l'ho lasciato sulla base dell'interfaccia proposta lezioni fa.

Eccetto il Database che sono costretto a fare più avanti ho provveduto a modificare il codice secondo quasi tutte le osservazioni fatte e mi scuso ancora per non aver sistemato a sufficienza in anticipo, come ho detto ho voluto fare troppo, non ho mandato in esaminazione molti file perchè erano ancor meno da code review di questi.

Commit:
Commit reflecting most of the corrections and suggestions by Luca Crema.

- Changed Names for classes
- Changed inappropriate variable and method names
- Added the few missing specs
- Removed flags from BroadcastReceiver
- Implemented TODOs
- shouldHandleIncomingSms() should have been package-private
- Fixed hardwiring for Intent.putExtras and Intent.getExtras

Signed-off-by: Riccardo De Zen <riccardo.dezen@studenti.unipd.it>

CremaLuca commented on lib/src/main/java/com/dezen/riccardo/smshandler/SMSManager.java:
```diff
@@ -9,77 +9,100 @@
+    private SMSHandler smsHandler;
+
+    private SMSManager(){
+        //prevent use of reflection to change constructor to public at runtime
```
Pazzesco che tu abbia pensato anche ad un possibile reflection, molto bravo

riccardo-crociani replied:
Bene, ho imparato cos'è un reflection

riccardo-crociani commented:
Con il secondo commit, dopo le correzzioni sugggerite si sono chiariti molti aspetti
Non ho ben capito come funziona sendImplicitBroadcast(Context context, Intent intent) e, di conseguenza perché viene invocato nel metodo onReceive(Context, Intent) della classe SMSReceiver 

- Ben formattato
- non ci sono più hardwire e TODO
- i nomi delle variabili sono abbastanza chiari

RicDeZen commented:
> Non ho ben capito come funziona sendImplicitBroadcast(Context context, Intent intent) e, di conseguenza perché viene invocato nel metodo onReceive(Context, Intent) della classe SMSReceiver

sendImplicitBroadcast vuole essere un metodo che recupera tutti i broadcastReceiver registrati dall'applicazione per inviargli dei broadcast direttamente, serve ad attivare broadcastReceiver in altri moduli, esterni alla libreria, in caso si voglia per esempio avviare forzatamente qualcosa mentre l'app è chiusa. Non basta fare Context.sendBroadcast() poichè la possibilità di inviare broadcast generali è stata bloccata in Android 8.0.

riccardo-crociani commented:
Beh geniale, ho capito grazie
